source/state_representation/src/robot/Jacobian.cpp

```diff
@@ -119,24 +202,55 @@ Jacobian Jacobian::copy() const {
   return result;
 }
 
-std::ostream& operator<<(std::ostream& os, const Jacobian& matrix) {
-  if (matrix.is_empty()) {
+std::ostream& operator<<(std::ostream& os, const Jacobian& jacobian) {
+  if (jacobian.is_empty()) {
     os << "Empty Jacobian";
   } else {
-    os << matrix.get_name() << " Jacobian" << std::endl;
+    os << jacobian.get_name() << " Jacobian associated to " << jacobian.frame_;
+    os << ", expressed in " << jacobian.reference_frame_ << std::endl;
     os << "joint names: [";
-    for (auto& n : matrix.get_joint_names()) { os << n << ", "; }
+    for (auto& n : jacobian.get_joint_names()) { os << n << ", "; }
     os << "]" << std::endl;
-    os << "number of rows: " << matrix.get_nb_rows() << std::endl;
-    os << "number of columns: " << matrix.get_nb_cols() << std::endl;
-    for (unsigned int i = 0; i < matrix.get_nb_rows(); ++i) {
-      os << "| " << matrix(i, 0);
-      for (unsigned int j = 1; j < matrix.get_nb_cols(); ++j) {
-        os << ", " << matrix(i, j);
+    os << "number of rows: " << jacobian.rows_ << std::endl;
+    os << "number of columns: " << jacobian.cols_ << std::endl;
+    for (unsigned int i = 0; i < jacobian.rows_; ++i) {
+      os << "| " << jacobian(i, 0);
+      for (unsigned int j = 1; j < jacobian.cols_; ++j) {
+        os << ", " << jacobian(i, j);
       }
       os << " |" << std::endl;
     }
   }
   return os;
 }
+
+Jacobian operator*(const CartesianPose& pose, const Jacobian& jacobian) {
+  // check compatibility
+  if (jacobian.is_empty()) {
+    throw EmptyStateException(jacobian.get_name() + " state is empty");
+  }
+  if (pose.is_empty()) {
+    throw EmptyStateException(pose.get_name() + " state is empty");
+  }
+  if (pose.get_name() != jacobian.get_reference_frame()) {
+    throw IncompatibleStatesException("The Jacobian and the input CartesianPose are incompatible, expected pose of "
+                                          + jacobian.get_reference_frame() + " got " + pose.get_name());
+  }
+  // number of rows of the jacobian should be 6 (incorrect if it has been transposed before)
+  if (jacobian.rows_ != 6) {
+    throw IncompatibleStatesException(
+        "The Jacobian and the input CartesianPose are incompatible, the Jacobian has probably been transposed before");
+  }
+  Jacobian result(jacobian);
+  // change the reference frame of all the columns
+  for (unsigned int i = 0; i < jacobian.cols_; ++i) {
+    // update position part
+    result.data_.col(i).head(3) = pose * jacobian.data_.col(i).head(3);
```

Comment:
You know what, I think this is actually wrong. The `pose * vector` operation is actually implemented as:
```
pose.get_orientation() * vector + pose.get_position();
```
Which is good if `vector` represents a `position`, but here I don't think it makes sense. I think it is actually enough to have:
```
result.data_.col(i).head(3) = pose.get_orientation() * jacobian.data_.col(i).head(3);
```

Reply:
Ok so I have found how Moveit is doing it. I guess it makes sense: https://github.com/ros-planning/moveit/blob/melodic-devel/moveit_core/robot_state/src/robot_state.cpp#L1250-L1254

Reply:
The syntax in that reference is very confusing. They initialise the jacobian to zeros, and then use a very roundabout way of `+=` to make the assignments.

For example, in the revolute joint, the angular part of the jacobian is updated as:
`jacobian.block<3, 1>(3, joint_index) = jacobian.block<3, 1>(3, joint_index) + joint_axis`
This makes it look like the transformation is actually "summing" the magnitude, but really it's just adding to zero. It should really just be a straight assignment.


But, I also think that in the reference they are computing the jacobian from zero for each link, so it's not really a direct comparison. We have a pre-computed jacobian relating the joint velocity to the twist of a given frame, and we want to move or rotate that frame. Rotating the frame should not change the magnitude of either the linear or angular velocity parts, simply rotating them as well. But, translating the frame should indeed change the magnitude of the linear part (and I believe would not change the angular velocity of the frame).

I feel like it might be reasonable to take the 6vector column and cast it as a CartesianTwist, then transform it with the CartesianPose and set the new twist as the column... but it would definitely need unit tests and a known (perhaps manually derived) case to check against.